File: telegram/src/main/java/by/mrrockka/service/TelegramPersonService.java
package by.mrrockka.service;

import by.mrrockka.domain.MessageMetadata;
import by.mrrockka.domain.Person;
import by.mrrockka.domain.TelegramPerson;
import by.mrrockka.mapper.TelegramPersonMapper;
import by.mrrockka.repo.person.TelegramPersonRepository;
import by.mrrockka.validation.mentions.PersonMentionsValidator;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Isolation;
import org.springframework.transaction.annotation.Propagation;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.UUID;
import java.util.stream.Stream;

@Service
@RequiredArgsConstructor
public class TelegramPersonService {

  private final TelegramPersonMapper telegramPersonMapper;
  private final PersonService personService;
  private final TelegramPersonRepository telegramPersonRepository;
  private final PersonMentionsValidator personMentionsValidator;

  @Transactional(isolation = Isolation.READ_COMMITTED, propagation = Propagation.REQUIRED)
  public List<TelegramPerson> storePersons(final MessageMetadata messageMetadata) {
    personMentionsValidator.validateMessageMentions(messageMetadata, 1);
    return storeMissed(messageMetadata);
  }

  @Transactional(isolation = Isolation.READ_COMMITTED, propagation = Propagation.REQUIRED)
  public TelegramPerson getByNicknameAndChatId(final String nickname, final Long chatId) {
    return telegramPersonRepository.findByNickname(chatId, nickname)
      .map(telegramPersonMapper::mapToTelegramPerson)
      .orElseGet(() -> saveNew(nickname, chatId));
  }

  public List<TelegramPerson> getAllByNicknamesAndChatId(final List<String> telegrams, final Long chatId) {
    return telegramPersonMapper.mapToTelegramPersons(
      telegramPersonRepository.findAllByChatIdAndNicknames(telegrams, chatId));
  }

  @Transactional(isolation = Isolation.READ_COMMITTED, propagation = Propagation.REQUIRED)
  public List<TelegramPerson> storeMissed(final MessageMetadata metadata) {
    final var nicknames = metadata.mentions().map(entity -> entity.text().replaceAll("@", "")).toList();
    final var newNicknames = personService.getNewNicknames(nicknames);
    final var notExistentInChat = telegramPersonRepository.findNotExistentInChat(nicknames, metadata.chatId());

    final var newPersons = newNicknames.stream()
      .map(nickname -> Person.personBuilder()
        .nickname(nickname)
        .id(UUID.randomUUID())
        .build())
      .toList();

    if (!newPersons.isEmpty()) {
      personService.storeAll(newPersons);
    }

    if (!notExistentInChat.isEmpty() || !newPersons.isEmpty()) {
      final var allNewPersonIds = Stream.of(
          newPersons.stream().map(Person::getId).toList(),
          notExistentInChat
        ).flatMap(List::stream)
        .toList();

      telegramPersonRepository.saveAll(allNewPersonIds, metadata.chatId());
    }

    return telegramPersonMapper.mapToTelegramPersons(
      telegramPersonRepository.findAllByChatIdAndNicknames(nicknames, metadata.chatId()));
  }

  private TelegramPerson saveNew(final String nickname, final Long chatId) {
    final var telegramPerson = TelegramPerson.telegramPersonBuilder()
      .id(UUID.randomUUID())
      .nickname(nickname)
      .chatId(chatId)
      .build();

    personService.store(telegramPerson);
    telegramPersonRepository.save(telegramPerson);
    return telegramPerson;
  }

}

File: telegram/src/main/java/by/mrrockka/service/TelegramPersonService.kt
package by.mrrockka.service

import by.mrrockka.domain.BasicPerson
import by.mrrockka.domain.MessageMetadata
import by.mrrockka.repo.ChatPersonsRepo
import by.mrrockka.repo.PersonRepo
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Propagation
import org.springframework.transaction.annotation.Transactional
import java.util.*

interface TelegramPersonService {
    fun findByMessage(metadata: MessageMetadata): List<BasicPerson>
    fun findOrAdd(metadata: MessageMetadata): List<BasicPerson>
    fun findOrAdd(nicknames: List<String>, chatId: Long): List<BasicPerson>
    fun findOrAdd(nickname: String, chatId: Long): BasicPerson
    fun findByFrom(metadata: MessageMetadata): BasicPerson
}

@Component
@Transactional(propagation = Propagation.REQUIRED)
open class TelegramPersonServiceImpl(
        private val personRepo: PersonRepo,
        private val chatPersonsRepo: ChatPersonsRepo,
) : TelegramPersonService {

    override fun findByMessage(metadata: MessageMetadata): List<BasicPerson> {
        val nicknames = metadata.mentions.map { it.text }
        return personRepo.findByNicknames(nicknames)
    }

    override fun findOrAdd(metadata: MessageMetadata): List<BasicPerson> = findOrAdd(metadata.mentions.map { it.text }, metadata.chatId)

    override fun findOrAdd(nicknames: List<String>, chatId: Long): List<BasicPerson> {
        val persons = personRepo.findByNicknames(nicknames)
        val newPersons = nicknames.newNicknames(persons)
                .let { nicknames ->
                    val newPersons = nicknames.map { nickname -> BasicPerson(nickname = nickname, id = UUID.randomUUID()) }
                    personRepo.store(newPersons)
                    newPersons
                }

        val allPersons = persons + newPersons
        chatPersonsRepo.store(allPersons.map { it.id }, chatId)

        return allPersons
    }

    override fun findOrAdd(nickname: String, chatId: Long): BasicPerson {
        return personRepo.findByNickname(nickname)
                .let { person ->
                    if (person != null) person
                    else {
                        val new = BasicPerson(nickname = nickname, id = UUID.randomUUID())
                        personRepo.store(new)
                        chatPersonsRepo.store(new.id, chatId)
                        new
                    }
                }
    }

    override fun findByFrom(metadata: MessageMetadata): BasicPerson {
        check(metadata.from?.username != null) { "User must have nickname to execute command" }
        val person = personRepo.findByNickname(metadata.from.username!!)
        check(person != null) { "Person does not found" }
        check(chatPersonsRepo.personChats(person.id).contains(metadata.chatId)) { "Person in not found in this chat" }
        return person
    }

    private fun List<String>.newNicknames(existing: List<BasicPerson>): List<String> {
        val existingNicknames = existing.map { it.nickname }
        return filter { !existingNicknames.contains(it) }
    }

}